server/on-demand-entry-handler.js

```diff
@@ -57,65 +66,79 @@ export default function onDemandEntryHandler (devMiddleware, compilers, {
         return addEntry(compilation, compiler.context, name, entry)
       })
 
-      Promise.all(allEntries)
-        .then(() => done())
-        .catch(done)
+      return Promise.all(allEntries)
     })
+  }
 
-    compiler.hooks.done.tap('NextJsOnDemandEntries', function (stats) {
-      // Wait until all the compilers mark the build as done.
-      currentBuilders.delete(compiler.name)
-      if (currentBuilders.size !== 0) return
+  multiCompiler.hooks.done.tap('NextJsOnDemandEntries', (multiStats) => {
+    const clientStats = multiStats.stats[0]
+    const { compilation } = clientStats
+    const hardFailedPages = compilation.errors
+      .filter(e => {
+        // Make sure to only pick errors which marked with missing modules
+        const hasNoModuleFoundError = /ENOENT/.test(e.message) || /Module not found/.test(e.message)
+        if (!hasNoModuleFoundError) return false
+
+        // The page itself is missing. So this is a failed page.
+        if (IS_BUNDLED_PAGE_REGEX.test(e.module.name)) return true
+
+        // No dependencies means this is a top level page.
+        // So this is a failed page.
+        return e.module.dependencies.length === 0
+      })
+      .map(e => e.module.chunks)
+      .reduce((a, b) => [...a, ...b], [])
+      .map(c => {
+        const pageName = ROUTE_NAME_REGEX.exec(c.name)[1]
+        return normalizePage(`/${pageName}`)
+      })
 
-      const { compilation } = stats
-      const hardFailedPages = compilation.errors
-        .filter(e => {
-          // Make sure to only pick errors which marked with missing modules
-          const hasNoModuleFoundError = /ENOENT/.test(e.message) || /Module not found/.test(e.message)
-          if (!hasNoModuleFoundError) return false
+    // compilation.entrypoints is a Map object, so iterating over it 0 is the key and 1 is the value
+    for (const [, entrypoint] of compilation.entrypoints.entries()) {
```

Comment:
The main change in this PR is that we're no longer calling `doneCallbacks` on `entries`, because this leads to timing issues. Instead we use webpack's `compilation.entrypoints` as the source of truth of what's compiled (which is logical, as it holds exactly what's compiled in this compilation run).